File: app/src/main/java/in/smslite/db/MessageDao.java
package in.smslite.db;

import android.arch.lifecycle.LiveData;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;
import android.database.Cursor;

import java.util.List;

/**
 * Created by rahul1993 on 11/11/2017.
 */
@Dao
public interface MessageDao {
  @Insert(onConflict = OnConflictStrategy.REPLACE)
  void insertMessage(Message message);

  @Query("Select * from Message group by address order by timestamp desc")
  LiveData<List<Message>> getMessage();

  @Query("Select * from Message Where address = :address order by timestamp asc")
  LiveData<List<Message>> getMessageListByAddress(String address);

  //  @Query("Select * from Message Where Category = :category group by address order by timestamp desc")
  @Query("select t1.* from message t1 Join (select address, MAX(timestamp) timestamp from message group by address)" +
      "t2 on t1.address = t2.address and t1.timestamp = t2.timestamp where category = :category order by timestamp desc")
  LiveData<List<Message>> getMessageListByCategory(int category);

  @Query("Select * from Message Where seen = 0 and Category = :category")
  Cursor getUnseenSmsCount(int category);

  @Query("Select * from Message Where seen = 0 and Category = :category order by timestamp desc")
  List<Message> getNotificationSummary(int category);

  @Query("Update Message Set seen = 1 Where Category = :category")
  void markAllSeen(int category);

  @Query("Update Message Set seen=1, read = 1 Where address = :address")
  void markAllRead(String address);

  @Query("select * from message where Type Like 2 order by timestamp desc")
  Cursor getSentSmsCount();

  @Query("select * from message where body Like \"%otp%\" group by address")
  List<Message> getOTPFOrTest();

  @Query("Update Message set type = 2 where timestamp LIKE :time")
  void updateSentSuccessful(Long time);

  @Query("Update Message set type = 4 where timestamp LIKE :time")
  void deliveredSmsSuccessfully(Long time);

  //  Query for failed sms
  @Query("Update Message set type = 5 where timestamp = :time")
  void updateSentFailedSms(Long time);

  @Query("Select body from Message where timestamp = :time")
  String getFailedSmsText(Long time);

  @Query("Delete from message where timestamp = :time")
  void deleteFailedMsg(Long time);

  //  Query for message search
//  @Query("Select * from Message where body LIKE :keyword or address LIKE :keyword group by address order by timestamp desc")
  @Query("select t1.* from message t1 Join (select address, MAX(timestamp) timestamp from message group by address)" +
      "t2 on t1.address = t2.address and t1.timestamp = t2.timestamp where body LIKE :keyword or t1.address LIKE :keyword" +
      " group by t1.address order by t1.timestamp desc")
  List<Message> searchMsg(String keyword);

  @Query("Delete from message where address = :address")
  void deleteSelectedConversation(String address);

  @Query("Delete from message where timestamp = :timestamp")
  void deleteSelectedMessage(Long timestamp);

//  Update the Category of message to blocked
  @Query("Update Message set category = :category where address Like :address")
  void moveToCategory(String address, int category);

//  Query to select all archive messages
  @Query("Select t1.* from message t1 JOIN (select address, Max(timestamp) timestamp from message where category = 6 group by address)" +
      "t2 on t1.address = t2.address AND t1.timestamp = t2.timestamp order by timestamp desc")
  LiveData<List<Message>> getArchiveMessage ();

  @Query("select t1.* from message t1 Inner Join (select address, Max(timestamp) timestamp from message where category = 5 group by address) t2" +
      " on t1.address = t2.address and t1.timestamp = t2.timestamp order by timestamp Desc")
  LiveData<List<Message>> getBlockedMessage();

//  @Query("Select * from Message Where widget = 1 order by timestamp desc" )
//  public List<Message> getWidgetMessage();
//
//  @Query("Update Message set widget = 1 where body like :name ")
//  public void updateWidgetMessage(String name);
}

File: app/src/main/java/in/smslite/db/MessageDao.kt
package `in`.smslite.db

import android.database.Cursor
import androidx.lifecycle.LiveData
import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query

/**
 * Created by rahul1993 on 11/11/2017.
 */
@Dao
interface MessageDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertMessage(message: Message?)

    @Query("Select * from Message group by address order by receivedDate and sentDate desc")
    fun getMessage(): LiveData<List<Message>>

    @Query("Select * from Message Where address = :address AND category=:category order by receivedDate and sentDate asc")
    fun getMessageListByAddress(address: String?, category: Int): LiveData<List<Message>>

    @Query("select t1.* from message t1 Join (select address, MAX(receivedDate) receivedDate from message where category =:category group by address)" +
            "t2 on t1.address = t2.address and t1.receivedDate = t2.receivedDate where category = :category order by receivedDate desc")
    fun getMessageListByCategory(category: Int?): LiveData<List<Message>>

    @Query("SELECT t1.* from Message t1 Join (SELECT address, MAX(receivedDate) receivedDate from Message where category = :category group by address) t2 on t1.address = t2.address and t1.receivedDate = t2.receivedDate where category = :category order by receivedDate desc")
    fun getMessageListByCategory1(category: Int?): LiveData<List<Message>>

    @Query("Select * from Message Where seen = 0 and Category = :category")
    fun getUnseenSmsCount(category: Int): Cursor?

    @Query("Select * from Message Where seen = 0 and Category = :category order by receivedDate and sentDate desc")
    suspend fun getNotificationSummary(category: Int): List<Message>

    @Query("Update Message Set seen = 1 Where Category = :category")
    suspend fun markAllSeen(category: Int?)

    @Query("Update Message Set seen=1, read = 1 Where address = :address")
    suspend fun markAllReadByAddress(address: String?)

    @Query("Update Message set Read = 1, seen = 1")
    suspend fun markAllRead()

    @Query("select * from message where Type Like 2 order by receivedDate and sentDate desc")
    fun getSentSmsCount(): Cursor?

    @get:Query("select * from message  where category = 4 order by receivedDate and sentDate desc")
    val oTPFOrTest: List<Message?>?

    @Query("Update Message set type = 2 where receivedDate and sentDate LIKE :time")
    suspend fun updateSentSuccessful(time: Long?)

    @Query("Update Message set type = 4 where receivedDate and sentDate LIKE :time")
    suspend fun deliveredSmsSuccessfully(time: Long?)

    //  Query for failed sms
    @Query("Update Message set type = 5 where receivedDate and sentDate = :time")
    suspend fun updateSentFailedSms(time: Long?)

    @Query("Select body from Message where receivedDate and sentDate = :time")
    suspend fun getFailedSmsText(time: Long?): String?

    @Query("Delete from message where receivedDate and sentDate = :time")
    suspend fun deleteFailedMsg(time: Long?)

    //  Query for message search
    //  @Query("Select * from Message where body LIKE :keyword or address LIKE :keyword group by address order by timestamp desc")
    @Query("select t1.* from message t1 Join (select address, MAX(receivedDate) receivedDate from message group by address)" +
            "t2 on t1.address = t2.address and t1.receivedDate = t2.receivedDate where body LIKE :keyword or t1.address LIKE :keyword" +
            " group by t1.address order by t1.receivedDate desc")
    fun searchMsg(keyword: String?): LiveData<List<Message>>

    @Query("Delete from message where address = :address")
    suspend fun deleteSelectedConversation(address: String?)

    @Query("Delete from message where  sentDate = :timestamp")
    suspend fun deleteSelectedMessage(timestamp: Long?)

    //  Update the Category of message to blocked
    @Query("Update Message set category = :category where address Like :address AND category=:presentCategory")
    suspend fun moveToCategory(address: String?, category: Int, presentCategory: Int)

    //  Query to select all archive messages
    @Query("Select t1.* from message t1 JOIN (select address, Max(receivedDate) receivedDate from message where category = 6 group by address)" +
            "t2 on t1.address = t2.address AND t1.receivedDate = t2.receivedDate order by receivedDate desc")
    fun getArchiveMessage(): LiveData<List<Message>>

    @get:Query("select t1.* from message t1 Inner Join (select address, Max(receivedDate) receivedDate from message where category = 5 group by address) t2" +
            " on t1.address = t2.address and t1.receivedDate = t2.receivedDate order by receivedDate Desc")
    val blockedMessage: LiveData<List<Message>>

    @get:Query("select * from message where category=1 order by receivedDate and sentDate desc")
    val primaryMessage: List<Message?>?

    //    @Query("Update message set sendFutureMessage = :value where address = :address")
//    suspend fun updateSendFutureMessage(address: String?, value: Int)
//
//    @Query("select t1.sendFutureMessage from message t1 join (select Min(receivedDate) receivedDate from message where address=:address) " +
//            "t2 on t1.receivedDate=t2.receivedDate where address = :address group by address")
//    suspend fun askSendFutureMessage(address: String?): Int
//
    @Query("select t1.futureCategory from message t1 Join(select Min(receivedDate) receivedDate from message where address=:address) " +
            "t2 on t1.receivedDate = t2.receivedDate where address = :address group by address")
    suspend fun findCategory(address: String?): Int

    @Query("update message set futureCategory =:category where address =:address")
    suspend fun updateFutureCategory(address: String?, category: Int)

}